src/esp/sensor/Base.h

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include "esp/io/json.h"
+#include "esp/scene/SceneNode.h"
+
+#include <string>
+
+namespace esp {
+namespace sensor {
+
+// Sensors need to extend this struct for their custom specification
+struct BaseSensorSpec {
+  // Specifies a unique idenfitier for the sensor for e.g. "rgb"
+  std::string UID = "base";
+  // Specifies the type of sensor
+  // TODO: Consider if this should be converted to an enum
+  std::string sensorType = "none";
+  ESP_SMART_POINTERS(BaseSensorSpec);
+};
+
+// Base class for a sensor register to the simulator
+// Sensor provides data from the environment to the agent
+class BaseSensor : public Magnum::SceneGraph::AbstractFeature3D {
+ public:
+  /**
+   * @brief Constructor, sets common private variable
+   * @param[in] node Instance of SceneNode to which sensor will be attached
+   * @param[in] spec Instance of BaseSensorSpec or derived structs describing
+   *                 specification of the sensor
+   */
+  explicit BaseSensor(scene::SceneNode& node, const BaseSensorSpec& spec);
+
+  /**
+   * @brief Provides the type of the sensor
+   * @return string type of the sensor
+   *
+   * NOTE: Override in child sensor class
+   */
+  virtual const std::string& getType() const = 0;
+
+  /**
+   * @brief Universally Unique Identifier for the current instance of the
+   *        sensor. Appends a unique string to UID defined by the
+   *        sensor specification.
+   * @return string UUID for the current instance
+   */
+  const std::string& getUUID() const;
+
+  /**
+   * @brief Get current observation for the sensor
+   * @return unique pointer to char observation
+   *
+   * NOTE: Override in child sensor class
+   */
+  virtual std::unique_ptr<char> getObservation() = 0;
+
+  /**
+   * @brief Assign current observation of the sensor to the
+   *        passed argument
+   * @param[in, out] obs Pointer to the variable to which current observation
+   *                     will be assigned
+   *
+   * NOTE: Override in child sensor class
+   */
+  virtual bool getObservation(void* obs);
+
+  /**
+   * @brief Get shape of the observation returned by the sensor
+   * @return vector<size_t> Array defining shape of the observation
+   *                        returned by the sensor for e.g. [2, 3]
+   *
+   * NOTE: Override in child sensor class
+   */
+  virtual const std::vector<size_t>& getObservationShape() const = 0;
+
+  /**
+   * @brief Get data type of the observation
+   * @return string Type of the observation returned by the sensor
+   *                For e.g. "uint32", "double".
+   *
+   * NOTE: Override in child class
+   */
+  virtual const std::string& getObservationDataType() const = 0;
+
+ protected:
+  ESP_SMART_POINTERS(BaseSensor);
+};
+
+// Factory for creating sensors
+class SensorFactory {
```

Comment:
One file for one class. It is easier for the developers to find the class by filenames.